File: frontend/src/types/serviceScale.ts
export type ScaleStatus = "DRAFT" | "PUBLISHED";
export type ScaleModality = "FT" | "ROCAM";

export interface ScaleCalendarDay {
  date: string;
  scale_id: number | null;
  title: string | null;
  status: ScaleStatus | null;
  team_count: number;
}

export interface ScaleCalendarResponse {
  year: number;
  month: number;
  days: ScaleCalendarDay[];
}

export interface StaffAbsenceFlag {
  kind: "FOLGA" | "FERIAS" | "LP";
  label: string;
}

export interface StaffRosterEntry {
  user_id: number;
  patente: string;
  nome_guerra: string;
  display_order: number;
  operational_rank: number;
  organizational_unit: "FIRST_PLATOON" | "SECOND_PLATOON" | "COMPANY_ADMIN";
  absences: StaffAbsenceFlag[];
}

export interface ScaleVehicleOption {
  id: number;
  prefixo: string;
  placa: string;
  modalidade: string;
}

export interface ScaleTeamMemberPublic {
  id: number;
  user_id: number;
  patente: string;
  nome_guerra: string;
  display_order: number;
  assigned_vehicle_id: number | null;
  assigned_vehicle_prefixo: string | null;
  role_label: string | null;
}

export interface ScaleTeamPublic {
  id: number;
  modality: ScaleModality;
  vehicle_id: number | null;
  vehicle_prefixo: string | null;
  vehicle_placa: string | null;
  start_datetime: string;
  end_datetime: string;
  mission_name: string;
  notes: string | null;
  members: ScaleTeamMemberPublic[];
}

export interface ServiceScalePublic {
  id: number;
  scale_date: string;
  title: string;
  description: string | null;
  fardamento: string | null;
  status: ScaleStatus;
  created_by_id: number;
  created_by_label: string | null;
  published_at: string | null;
  current_version_number: number | null;
  created_at: string;
  updated_at: string;
  teams: ScaleTeamPublic[];
}

export interface ScaleVersionPublic {
  id: number;
  service_scale_id: number;
  version_number: number;
  published_at: string;
  published_by_id: number;
  published_by_label: string | null;
  change_summary: string | null;
  dejem_integrated_count: number;
  created_at: string;
}

export interface ScaleVersionDetail extends ScaleVersionPublic {
  export_text: string;
}

export interface ScaleDayDetailResponse {
  scale: ServiceScalePublic | null;
  staff_roster: StaffRosterEntry[];
  vehicles_ft: ScaleVehicleOption[];
  vehicles_ro_cam: ScaleVehicleOption[];
  dejem_blocks: DejemMapBlock[];
}

export interface DejemMapMember {
  user_id: number;
  patente: string;
  nome_guerra: string;
  display_order: number;
}

export interface DejemMapBlock {
  shift_id: number;
  title: string;
  shift_type: "FT" | "ROCAM" | "OUTROS";
  start_time: string;
  end_time: string;
  status: string;
  vehicle_prefixo: string | null;
  members: DejemMapMember[];
}

export interface ScaleLogFeedItem {
  id: number;
  service_scale_id: number;
  scale_date: string;
  scale_title: string;
  action_type: string;
  description: string;
  created_at: string;
  actor_label: string;
}

export interface ScaleHistoryEntry {
  id: number;
  scale_date: string;
  title: string;
  status: ScaleStatus;
  team_count: number;
  published_at: string | null;
  updated_at: string;
}

export interface ScaleHistoryResponse {
  items: ScaleHistoryEntry[];
  total: number;
}

export interface ScaleTeamMemberInput {
  user_id: number;
  assigned_vehicle_id?: number | null;
  role_label?: string | null;
}

export interface ScaleTeamCreatePayload {
  modality: ScaleModality;
  vehicle_id?: number | null;
  start_datetime: string;
  end_datetime: string;
  mission_name: string;
  notes?: string | null;
  members: ScaleTeamMemberInput[];
}

export interface ScaleTeamUpdatePayload {
  modality?: ScaleModality;
  vehicle_id?: number | null;
  start_datetime?: string;
  end_datetime?: string;
  mission_name?: string;
  notes?: string | null;
  members?: ScaleTeamMemberInput[];
}

export const FT_MISSION_PRESETS = ["Tático Comando", "Supervisor Tático", "Força Tática"] as const;
export const ROCAM_MISSION_PRESETS = ["ROCAM 1", "ROCAM 2", "ROCAM 3"] as const;

/** Funções fixas — ordem = ordem da mensagem operacional. */
export const FT_TEAM_ROLES = [
  "Comandante da Equipe",
  "Motorista",
  "3º Homem",
  "4º Homem",
] as const;

export const ROCAM_TEAM_ROLES = ["Comandante da Equipe", "Moto 2", "Moto 3"] as const;

export type FtTeamRole = (typeof FT_TEAM_ROLES)[number];
export type RocamTeamRole = (typeof ROCAM_TEAM_ROLES)[number];
export type TeamRole = FtTeamRole | RocamTeamRole;

export function teamRolesFor(modality: ScaleModality): readonly TeamRole[] {
  return modality === "ROCAM" ? ROCAM_TEAM_ROLES : FT_TEAM_ROLES;
}

export function sortMembersByRole<T extends { role_label?: string | null; display_order?: number; nome_guerra?: string }>(
  modality: ScaleModality,
  members: T[],
): T[] {
  const order = new Map(teamRolesFor(modality).map((r, i) => [r, i]));
  return [...members].sort((a, b) => {
    const ai = a.role_label != null ? order.get(a.role_label as TeamRole) : undefined;
    const bi = b.role_label != null ? order.get(b.role_label as TeamRole) : undefined;
    if (ai != null && bi != null) return ai - bi;
    if (ai != null) return -1;
    if (bi != null) return 1;
    const ao = a.display_order ?? 0;
    const bo = b.display_order ?? 0;
    if (ao !== bo) return ao - bo;
    return (a.nome_guerra ?? "").localeCompare(b.nome_guerra ?? "", "pt-BR");
  });
}

export interface ScaleExportResponse {
  text: string;
}

File: backend/schemas/service_scale.py
from datetime import date, datetime

from pydantic import BaseModel, Field, model_validator

from models.service_scale import ScaleLogAction, ScaleModality, ScaleStatus
from models.user import OrganizationalUnit
from schemas.dejem import DejemMapBlock

FT_MISSION_PRESETS = ("Tático Comando", "Supervisor Tático", "Força Tática")
ROCAM_MISSION_PRESETS = ("ROCAM 1", "ROCAM 2", "ROCAM 3")

MAX_FT_MEMBERS = 4
MAX_ROCAM_MEMBERS = 3


class ScaleTeamMemberInput(BaseModel):
    user_id: int
    assigned_vehicle_id: int | None = None
    role_label: str | None = Field(default=None, max_length=64)


class ScaleTeamMemberPublic(BaseModel):
    id: int
    user_id: int
    patente: str
    nome_guerra: str
    display_order: int
    assigned_vehicle_id: int | None
    assigned_vehicle_prefixo: str | None
    role_label: str | None

    model_config = {"from_attributes": False}


class ScaleTeamPublic(BaseModel):
    id: int
    modality: ScaleModality
    vehicle_id: int | None
    vehicle_prefixo: str | None
    vehicle_placa: str | None
    start_datetime: datetime
    end_datetime: datetime
    mission_name: str
    notes: str | None
    members: list[ScaleTeamMemberPublic]

    model_config = {"from_attributes": False}


class ServiceScaleCreate(BaseModel):
    scale_date: date
    title: str = Field(..., min_length=1, max_length=256)
    description: str | None = Field(default=None, max_length=4000)
    fardamento: str | None = Field(default=None, max_length=256)
    status: ScaleStatus = ScaleStatus.DRAFT


class ServiceScaleUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=256)
    description: str | None = Field(default=None, max_length=4000)
    fardamento: str | None = Field(default=None, max_length=256)


class ScaleTeamCreate(BaseModel):
    modality: ScaleModality
    vehicle_id: int | None = None
    start_datetime: datetime
    end_datetime: datetime
    mission_name: str = Field(..., min_length=1, max_length=256)
    notes: str | None = Field(default=None, max_length=4000)
    members: list[ScaleTeamMemberInput] = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_range(self) -> "ScaleTeamCreate":
        if self.end_datetime <= self.start_datetime:
            raise ValueError("Horário final deve ser posterior ao inicial")
        if self.modality == ScaleModality.ROCAM:
            if self.vehicle_id is not None:
                raise ValueError("Equipe ROCAM não possui viatura principal")
            for m in self.members:
                if not m.assigned_vehicle_id:
                    raise ValueError("Cada policial ROCAM deve ter uma moto vinculada")
        elif self.modality == ScaleModality.FT and not self.vehicle_id:
            raise ValueError("Viatura FT é obrigatória")
        return self


class ScaleTeamUpdate(BaseModel):
    modality: ScaleModality | None = None
    vehicle_id: int | None = None
    start_datetime: datetime | None = None
    end_datetime: datetime | None = None
    mission_name: str | None = Field(default=None, min_length=1, max_length=256)
    notes: str | None = Field(default=None, max_length=4000)
    members: list[ScaleTeamMemberInput] | None = None

    @model_validator(mode="after")
    def validate_modality_rules(self) -> "ScaleTeamUpdate":
        if self.members is None:
            return self
        if self.modality == ScaleModality.ROCAM:
            for m in self.members:
                if not m.assigned_vehicle_id:
                    raise ValueError("Cada policial ROCAM deve ter uma moto vinculada")
        return self


class ScaleTeamMembersUpdate(BaseModel):
    members: list[ScaleTeamMemberInput]

    @model_validator(mode="after")
    def validate_members(self) -> "ScaleTeamMembersUpdate":
        if not self.members:
            raise ValueError("Informe ao menos um policial")
        return self


class ServiceScalePublic(BaseModel):
    id: int
    scale_date: date
    title: str
    description: str | None
    fardamento: str | None = None
    status: ScaleStatus
    created_by_id: int
    created_by_label: str | None = None
    published_at: datetime | None
    current_version_number: int | None = None
    created_at: datetime
    updated_at: datetime
    teams: list[ScaleTeamPublic]

    model_config = {"from_attributes": False}


class ScaleVersionPublic(BaseModel):
    id: int
    service_scale_id: int
    version_number: int
    published_at: datetime
    published_by_id: int
    published_by_label: str | None = None
    change_summary: str | None = None
    dejem_integrated_count: int = 0
    created_at: datetime


class ScaleVersionDetail(ScaleVersionPublic):
    export_text: str


class ScaleCalendarDay(BaseModel):
    date: date
    scale_id: int | None
    title: str | None
    status: ScaleStatus | None
    team_count: int


class ScaleCalendarResponse(BaseModel):
    year: int
    month: int
    days: list[ScaleCalendarDay]


class StaffAbsenceFlag(BaseModel):
    kind: str
    label: str


class StaffRosterEntry(BaseModel):
    user_id: int
    patente: str
    nome_guerra: str
    display_order: int
    operational_rank: int
    organizational_unit: OrganizationalUnit
    absences: list[StaffAbsenceFlag]


class ScaleVehicleOption(BaseModel):
    id: int
    prefixo: str
    placa: str
    modalidade: str


class ScaleDayDetailResponse(BaseModel):
    scale: ServiceScalePublic | None
    staff_roster: list[StaffRosterEntry]
    vehicles_ft: list[ScaleVehicleOption]
    vehicles_ro_cam: list[ScaleVehicleOption]
    dejem_blocks: list[DejemMapBlock] = Field(default_factory=list)


class ScaleLogFeedItem(BaseModel):
    id: int
    service_scale_id: int
    scale_date: date
    scale_title: str
    action_type: ScaleLogAction
    description: str
    created_at: datetime
    actor_label: str

    model_config = {"from_attributes": False}


class ScaleHistoryEntry(BaseModel):
    id: int
    scale_date: date
    title: str
    status: ScaleStatus
    team_count: int
    published_at: datetime | None
    updated_at: datetime

    model_config = {"from_attributes": False}


class ScaleHistoryResponse(BaseModel):
    items: list[ScaleHistoryEntry]
    total: int


class ScaleExportResponse(BaseModel):
    text: str


class ScalePublishPreviewRequest(BaseModel):
    description: str | None = Field(default=None, max_length=4000)


class ScalePublishPreviewResponse(BaseModel):
    text: str
    fardamento: str | None = None
    description: str | None = None
    team_count: int = 0
    dejem_count: int = 0


class ScaleMessageTemplatePublic(BaseModel):
    id: int
    slug: str
    name: str
    body_text: str
    is_default: bool
    is_active: bool

    model_config = {"from_attributes": True}
